File: client/ts/src/framer/iterator.ts
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { type Stream, type WebSocketClient } from "@synnaxlabs/freighter";
import {
  type CrudeTimeRange,
  type CrudeTimeSpan,
  type CrudeTimeStamp,
  TimeRange,
  TimeSpan,
  TimeStamp,
} from "@synnaxlabs/x";

import { type channel } from "@/channel";
import { ReadAdapter } from "@/framer/adapter";
import { WSIteratorCodec } from "@/framer/codec";
import { Frame } from "@/framer/frame";
import {
  IteratorCommand,
  type IteratorRequest,
  iteratorReqZ,
  IteratorResponseVariant,
  iteratorResZ,
} from "@/framer/iterator.types";
import { StreamProxy } from "@/framer/streamProxy";

export {
  IteratorCommand,
  type IteratorRequest,
  type IteratorResponse,
  IteratorResponseVariant,
  iteratorResZ,
} from "@/framer/iterator.types";

export const AUTO_SPAN = new TimeSpan(-1);

export interface IteratorConfig {
  /** chunkSize is the maximum number of samples contained per channel in the frame
   * resulting from a call to next with {@link AUTO_SPAN}.
   */
  chunkSize?: number;
  /**
   * downsampleFactor is the factor to downsample the data by. If downsampleFactor is
   * less than or equal to 1, no downsampling will be performed.
   */
  downsampleFactor?: number;
}

/**
 * Used to iterate over a clusters telemetry in time-order. It should not be
 * instantiated directly, and should instead be instantiated via the SegmentClient.
 *
 * Using an iterator is ideal when querying/processing large ranges of data, but
 * is relatively complex and difficult to use. If you're looking to retrieve
 *  telemetry between two timestamps, see the SegmentClient.read method.
 */
export class Iterator {
  private readonly stream: StreamProxy<typeof iteratorReqZ, typeof iteratorResZ>;
  private readonly adapter: ReadAdapter;
  value: Frame;

  private constructor(
    stream: Stream<typeof iteratorReqZ, typeof iteratorResZ>,
    adapter: ReadAdapter,
  ) {
    this.stream = new StreamProxy("Iterator", stream);
    this.value = new Frame();
    this.adapter = adapter;
  }

  /**
   * Opens the iterator, configuring it to iterate over the telemetry in the
   * channels with the given keys within the provided time range.
   *
   * @param tr - The time range to iterate over.
   * @param channels - The channels for the iterator to iterate over (can be provided
   * in keys or names).
   * @param retriever - Retriever used to retrieve channel keys from names.
   * @param client - The stream client allowing streaming of iterated data.
   * @param opts - See {@link IteratorConfig}.
   */
  static async _open(
    tr: CrudeTimeRange,
    channels: channel.Params,
    retriever: channel.Retriever,
    client: WebSocketClient,
    opts: IteratorConfig = {},
  ): Promise<Iterator> {
    const adapter = await ReadAdapter.open(retriever, channels);
    client = client.withCodec(new WSIteratorCodec(adapter.codec));
    const stream = await client.stream("/frame/iterate", iteratorReqZ, iteratorResZ);
    const iter = new Iterator(stream, adapter);
    await iter.execute({
      command: IteratorCommand.Open,
      keys: Array.from(adapter.keys),
      bounds: new TimeRange(tr),
      chunkSize: opts.chunkSize ?? 1e5,
      downsampleFactor: opts.downsampleFactor ?? 1,
    });
    return iter;
  }

  /**
   * Reads the next time span of telemetry for each channel in the iterator.
   *
   * @param span - The time span to read. A negative span is equivalent
   * to calling prev with the absolute value of the span. If the span is
   * AUTO_SPAN, the iterator will automatically determine the span to read.
   * This is useful for iterating over an entire range efficiently.
   *
   * @returns false if a segment satisfying the request can't be found for a
   * particular channel or the iterator has accumulated an error.
   */
  async next(span: CrudeTimeSpan = AUTO_SPAN): Promise<boolean> {
    return await this.execute({
      command: IteratorCommand.Next,
      span: new TimeSpan(span),
    });
  }

  /**
   * Reads the previous time span of telemetry for each channel in the iterator.
   *
   * @param span - The time span to read. A negative span is equivalent
   * to calling next with the absolute value of the span. If the span is
   * AUTO_SPAN, the iterator will automatically determine the span to read.
   * This is useful for iterating over an entire range efficiently.
   *
   * @returns false if a segment satisfying the request can't be found for a particular
   * channel or the iterator has accumulated an error.
   */
  async prev(span: CrudeTimeSpan = AUTO_SPAN): Promise<boolean> {
    return await this.execute({
      command: IteratorCommand.Prev,
      span: new TimeSpan(span),
    });
  }

  /**
   * Seeks the iterator to the first segment in the time range, but does not read
   * it. Also invalidates the iterator. The iterator will not be considered valid
   * until a call to next or prev.
   *
   * @returns false if the iterator is not pointing to a valid segment for a particular
   * channel or has accumulated an error.
   */
  async seekFirst(): Promise<boolean> {
    return await this.execute({ command: IteratorCommand.SeekFirst });
  }

  /** Seeks the iterator to the last segment in the time range, but does not read it.
   * Also invalidates the iterator. The iterator will not be considered valid
   * until a call to next or prev.
   *
   * @returns false if the iterator is not pointing to a valid segment for a particular
   * channel or has accumulated an error.
   */
  async seekLast(): Promise<boolean> {
    return await this.execute({ command: IteratorCommand.SeekLast });
  }

  /**
   * Seeks the iterator to the first segment whose start is less than or equal to
   * the provided timestamp. Also invalidates the iterator. The iterator will not be
   * considered valid until a call to next or prev.
   *
   * @returns false if the iterator is not pointing to a valid segment for a particular
   * channel or has accumulated an error.
   */
  async seekLE(stamp: CrudeTimeStamp): Promise<boolean> {
    return await this.execute({
      command: IteratorCommand.SeekLE,
      stamp: new TimeStamp(stamp),
    });
  }

  /**
   * Seeks the iterator to the first segment whose start is greater than or equal to
   * the provided timestamp. Also invalidates the iterator. The iterator will not be
   * considered valid until a call to next or prev.
   *
   * @returns false if the iterator is not pointing to a valid segment for a particular
   * channel or has accumulated an error.
   */
  async seekGE(stamp: CrudeTimeStamp): Promise<boolean> {
    return await this.execute({
      command: IteratorCommand.SeekGE,
      stamp: new TimeStamp(stamp),
    });
  }

  /**
   * @returns true if the iterator value contains a valid segment, and false otherwise.
   * valid most commonly returns false when the iterator is exhausted or has accumulated
   * an error.
   */
  async valid(): Promise<boolean> {
    return await this.execute({ command: IteratorCommand.Valid });
  }

  /**
   * Closes the iterator. An iterator MUST be closed after use, and this method
   * should probably be placed in a 'finally' block. If the iterator is not closed,
   * it may leak resources.
   */
  async close(): Promise<void> {
    await this.stream.closeAndAck();
  }

  [Symbol.asyncIterator](): AsyncIterator<Frame, any, undefined> {
    return new IteratorIterator(this);
  }

  private async execute(request: IteratorRequest): Promise<boolean> {
    this.stream.send(request);
    this.value = new Frame();
    while (true) {
      const res = await this.stream.receive();
      if (res.variant === IteratorResponseVariant.Ack) return res.ack;
      this.value.push(this.adapter.adapt(new Frame(res.frame)));
    }
  }
}

class IteratorIterator implements AsyncIterator<Frame> {
  private readonly iter: Iterator;
  private open: boolean = false;

  constructor(iter: Iterator) {
    this.iter = iter;
  }

  async next(): Promise<IteratorResult<Frame, any>> {
    try {
      let ok = true;
      if (!this.open) {
        if (!(await this.iter.seekFirst())) ok = false;
        this.open = true;
      }
      if (!(await this.iter.next())) ok = false;
      if (!ok) await this.iter.close();
      return { done: !ok, value: this.iter.value };
    } catch (e) {
      await this.iter.close();
      throw e;
    }
  }
}

File: client/go/framer/iterator.go
package framer

import (
	"context"
	"github.com/synnaxlabs/freighter"
	"github.com/synnaxlabs/synnax/pkg/api"
	"github.com/synnaxlabs/synnax/pkg/distribution/channel"
	"github.com/synnaxlabs/synnax/pkg/distribution/framer/core"
	"github.com/synnaxlabs/synnax/pkg/distribution/framer/iterator"
	"github.com/synnaxlabs/x/errors"
	"github.com/synnaxlabs/x/telem"
)

const AutoSpan = api.FrameIteratorAutoSpan

type (
	IteratorRequest  = api.FrameIteratorRequest
	IteratorResponse = api.FrameIteratorResponse
	IteratorStream   = freighter.ClientStream[IteratorRequest, IteratorResponse]
	IteratorClient   = freighter.StreamClient[IteratorRequest, IteratorResponse]
	IteratorConfig   struct {
		Keys      channel.Keys
		Bounds    telem.TimeRange
		ChunkSize int64
	}
)

type Iterator struct {
	stream IteratorStream
	value  api.Frame
}

func openIterator(ctx context.Context, client IteratorClient, cfg IteratorConfig) (*Iterator, error) {
	s, err := client.Stream(ctx, "")
	if err != nil {
		return nil, err
	}
	i := &Iterator{stream: s}
	_, err = i.exec(ctx, IteratorRequest{Bounds: cfg.Bounds, ChunkSize: cfg.ChunkSize, Keys: cfg.Keys})
	return i, err
}

func (i *Iterator) Next(ctx context.Context, span telem.TimeSpan) bool {
	v, _ := i.exec(ctx, IteratorRequest{Command: iterator.Next, Span: span})
	return v
}

func (i *Iterator) Prev(ctx context.Context, span telem.TimeSpan) bool {
	v, _ := i.exec(ctx, IteratorRequest{Command: iterator.Prev, Span: span})
	return v
}

func (i *Iterator) SeekFirst(ctx context.Context) bool {
	v, _ := i.exec(ctx, IteratorRequest{Command: iterator.SeekFirst})
	return v
}

func (i *Iterator) SeekLE(ctx context.Context, stamp telem.TimeStamp) bool {
	v, _ := i.exec(ctx, IteratorRequest{Command: iterator.SeekLE, Stamp: stamp})
	return v
}

func (i *Iterator) SeekGE(ctx context.Context, stamp telem.TimeStamp) bool {
	v, _ := i.exec(ctx, IteratorRequest{Command: iterator.SeekGE, Stamp: stamp})
	return v
}

func (i *Iterator) Value() core.Frame {
	return i.value
}

func (i *Iterator) Valid() bool {
	v, _ := i.exec(context.Background(), IteratorRequest{Command: iterator.Valid})
	return v
}

func (i *Iterator) Close() error {
	if err := i.stream.CloseSend(); err != nil {
		return err
	}
	_, err := i.stream.Receive()
	return errors.Skip(err, freighter.EOF)
}

func (i *Iterator) exec(ctx context.Context, req IteratorRequest) (bool, error) {
	if err := i.stream.Send(req); err != nil {
		return false, err
	}
	i.value = core.Frame{}
	for {
		res, err := i.stream.Receive()
		if err != nil {
			return false, err
		}
		if res.Variant == iterator.AckResponse {
			return res.Ack, nil
		}
		i.value = i.value.Extend(res.Frame)
	}
}
